data_structures/skip_list.cpp

```diff
@@ -0,0 +1,207 @@
+using std::vector;
+using std::endl;
+
+#define MAXLVL 2   ///< maximum level of skip list
+#define P 0.5      ///< current probability for "coin toss"
+  
+
+/**
+ *  Node structure [Key][Node*, Node*...]
+*/
+struct Node { 
+    int key; 
+    void* value;
+    /*Forward Array*/
+    vector<Node*> forward;
+    Node(int key, int level, void* value); 
+};
+
+  
+Node::Node(int key, int level, void* value) { 
+    this->key = key; 
+  
+    /*Initialization of forward vector*/
+    for (int i = 0; i < (level+1); i++){
+        forward.push_back(NULL);
+    }
+} 
+  
+// Class for Skip list 
+class SkipList { 
+    int level; 
+    Node *header; 
+public: 
+    SkipList(); 
+    int randomLevel(); 
+    void insertElement(int, void*); 
+    void deleteElement(int); 
+    void* searchElement(int); 
+    void displayList(); 
+    ~SkipList();
+};
+
+
+  
+SkipList::SkipList() { 
+    level = 0; 
+    /* Header initialization*/
+    header = new Node(-1, MAXLVL, NULL); 
+} 
+
+
+SkipList::~SkipList(){
+    for(int i=0; i <= level; i++) { 
+        Node *node = header->forward[i]; 
+        Node* temp;
+        while(node != NULL) { 
+            temp = node;
+            node = node->forward[i];
+            delete temp;
+        } 
+    } 
+    delete header;
+}
+
+  
+/**
+ * Returns random level for skip list;
+*/
+int SkipList::randomLevel() { 
+    int lvl = 0; 
+    while(static_cast<float>(std::rand())/RAND_MAX < P && lvl < MAXLVL) lvl++;  
+    return lvl; 
+} 
+  
+  
+// Insert given key in skip list 
+void SkipList::insertElement(int key, void* value) { 
+    std::cout << "Inserting" << key << "...";
+    Node *x = header; 
+    Node *update[MAXLVL+1]; 
+    memset(update, 0, sizeof(Node*)*(MAXLVL+1)); 
+
+
+    for(int i = level; i >= 0; i--) { 
+        while(x->forward[i] != NULL && x->forward[i]->key < key)  x = x->forward[i]; 
+        update[i] = x; 
+    } 
+
+    x = x->forward[0]; 
+
+    bool doesnt_exist = (x == NULL || x->key != key);
+    if (doesnt_exist) { 
+        int rlevel = randomLevel(); 
+  
+        if(rlevel > level) { 
+            for(int i=level+1;i<rlevel+1;i++) update[i] = header; 
+
+            /*Update current level*/
+            level = rlevel; 
+        } 
+  
+        Node* n = new Node(key, rlevel, value);
+        for(int i=0;i<=rlevel;i++) { 
+            n->forward[i] = update[i]->forward[i]; 
+            update[i]->forward[i] = n; 
+        } 
+        std::cout << "Inserted" << endl;
+
+    } else {
+        std::cout << "Exists" << endl;
+    }
+} 
+  
+  /**Delete document by key*/
+void SkipList::deleteElement(int key) 
+{ 
+    Node *x = header; 
+  
+    Node *update[MAXLVL+1]; 
+    memset(update, 0, sizeof(Node*)*(MAXLVL+1)); 
+  
+    for(int i = level; i >= 0; i--)  { 
+        while(x->forward[i] != NULL  &&  x->forward[i]->key < key)  x = x->forward[i]; 
+        update[i] = x; 
+    } 
+  
+    x = x->forward[0]; 
+  
+    bool doesnt_exist = (x == NULL || x->key != key);
```

Comment:
in C++, NULL is deprecated and `nullptr` should be used instead. Please update wherever being used.

Reply:
alright